src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,291 @@
+        /// and only fetch the model once for every batch. The bool parameter is unused.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<bool, DocumentSymbolDataModel?> _computeDataModelQueue;
+
+        /// <summary>
+        /// Queue to batch up work to do to filter and sort the data model. The bool parameter is unused.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<bool, DocumentSymbolDataModel?> _filterAndSortDataModelQueue;
+
+        /// <summary>
+        /// Queue to batch up work to do to highlight the currently selected symbol node, expand/collapse nodes,
+        /// then update the UI.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<ExpansionOption> _highlightExpandAndPresentItemsQueue;
+
+        /// <summary>
+        /// Keeps track of the current primary and secondary text views. Should only be accessed by the UI thread.
+        /// </summary>
+        private readonly Dictionary<IVsTextView, ITextView> _trackedTextViews = new();
+
+        public DocumentOutlineControl(
+            ILanguageServiceBrokerShim languageServiceBroker,
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListener asyncListener,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IVsCodeWindow codeWindow)
+        {
+            InitializeComponent();
+
+            _languageServiceBroker = languageServiceBroker;
+            _threadingContext = threadingContext;
+            _asyncListener = asyncListener;
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            _codeWindow = codeWindow;
+            _cancellationTokenSource = new CancellationTokenSource();
+            _cancellationToken = _cancellationTokenSource.Token;
+            SortOption = SortOption.Location;
+
+            _computeDataModelQueue = new AsyncBatchingWorkQueue<bool, DocumentSymbolDataModel?>(
+                DelayTimeSpan.Short,
+                ComputeDataModelAsync,
+                EqualityComparer<bool>.Default,
+                asyncListener,
+                _cancellationToken);
+
+            _filterAndSortDataModelQueue = new AsyncBatchingWorkQueue<bool, DocumentSymbolDataModel?>(
+                DelayTimeSpan.NearImmediate,
+                FilterAndSortDataModelAsync,
+                EqualityComparer<bool>.Default,
+                asyncListener,
+                _cancellationToken);
+
+            _highlightExpandAndPresentItemsQueue = new AsyncBatchingWorkQueue<ExpansionOption>(
+                DelayTimeSpan.NearImmediate,
+                HighlightExpandAndPresentItemsAsync,
+                asyncListener,
+                _cancellationToken);
+
+            // We don't think the shell is initialized lazily, so we'll Debug.Fail(), but if it was we'd still
+            // see the view created later so this will still function.
```

Comment:
@jasonmalinowski @emilyanas2323 i don't think this is trye.  say that GetPrimaryView fails, well then 'primaryTextView' will be null right?  And that will immediately AV in `_trackedTextViews[primaryTextView]` since dictionaries do not allow null keys.

Reply:
Correct: the text buffer should be coming directly from the IVsCodeWindow or something.

Reply:
Let's defer this until the navbar code refactoring :)